[BUG] Filter on dates with subquery results in ArrayIndexOutOfBoundsException

**Describe the bug**

I was writing some test code and ran into this failure:

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
  at ai.rapids.cudf.Table.<init>(Table.java:58)
  at com.nvidia.spark.rapids.GpuColumnVector.from(GpuColumnVector.java:638)
  at com.nvidia.spark.rapids.GpuFilter$.$anonfun$doFilter$2(basicPhysicalOperators.scala:505)
  at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:29)
  at com.nvidia.spark.rapids.GpuFilter$.$anonfun$doFilter$1(basicPhysicalOperators.scala:503)
  at scala.Option.map(Option.scala:230)
  at com.nvidia.spark.rapids.GpuFilter$.doFilter(basicPhysicalOperators.scala:502)
  at com.nvidia.spark.rapids.GpuFilter$.apply(basicPhysicalOperators.scala:554)
  at com.nvidia.spark.rapids.GpuFilter$.$anonfun$filterAndCloseWithRetry$3(basicPhysicalOperators.scala:479)
  at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:29)
  at com.nvidia.spark.rapids.GpuFilter$.$anonfun$filterAndCloseWithRetry$2(basicPhysicalOperators.scala:478)
  at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:29)
  at com.nvidia.spark.rapids.GpuFilter$.$anonfun$filterAndCloseWithRetry$1(basicPhysicalOperators.scala:477)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$AutoCloseableAttemptSpliterator.next(RmmRapidsRetryIterator.scala:424)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:528)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:458)
  at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
  at scala.collection.Iterator$$anon$11.next(Iterator.scala:496)
  at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
  at scala.collection.Iterator.foreach(Iterator.scala:943)
  at scala.collection.Iterator.foreach$(Iterator.scala:943)
  at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
  at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:199)
  at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:192)
  at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1431)
  at scala.collection.TraversableOnce.sum(TraversableOnce.scala:262)
  at scala.collection.TraversableOnce.sum$(TraversableOnce.scala:262)
  at scala.collection.AbstractIterator.sum(Iterator.scala:1431)
  at com.nvidia.spark.rapids.GpuCoalesceBatches.$anonfun$internalDoExecuteColumnar$1(GpuCoalesceBatches.scala:872)
  at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:855)
```

To reproduce in scala:

```
val df = Seq((1)).toDF("seq")
df.write.parquet("testsubquery")
val df = spark.read.parquet("testsubquery")
val curDate = df.withColumn("current_date",current_date().as("current_date"))
curDate.createOrReplaceTempView("curDate")
spark.sql("select current_date, ((select last(current_date) from curDate) + interval 1 day) as test from curDate").createOrReplaceTempView("test2")
spark.sql("select * from test2 where test < current_date").show()
```

This works fine on CPU side and it works fine with > because there is an actual row there.

Result on CPU:
```
scala> spark.sql("select * from test2 where test < current_date").show()
+------------+----+
|current_date|test|
+------------+----+
+------------+----+
```

--- a/integration_tests/src/main/python/cmp_test.py
+++ b/integration_tests/src/main/python/cmp_test.py
@@ -1,4 +1,4 @@
-# Copyright (c) 2020-2022, NVIDIA CORPORATION.
+# Copyright (c) 2020-2023, NVIDIA CORPORATION.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.

--- a/integration_tests/src/main/python/cmp_test.py
+++ b/integration_tests/src/main/python/cmp_test.py
@@ -272,6 +272,27 @@ def test_filter_with_project(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
             lambda spark : two_col_df(spark, BooleanGen(), data_gen).filter(f.col('a')).selectExpr('*', 'a as a2'))
 
+# It takes quite a bit to get filter to have a column it can filter on, but
+# no columns to actually filter. We are making it happen here with a sub-query
+# and some constants that then make it so all we need is the number of rows
+# of input.
+@pytest.mark.parametrize('op', ['>', '<'])
+def test_empty_filter(op, spark_tmp_path):
+
+    def do_it(spark):
+        df = spark.createDataFrame([(14, "Tom"), (23, "Alice"), (16, "Bob")], ["age", "name"])
+        # we repartition the data to 3 because for some reason spark writes 4 files for 3 rows.
+        # In this case that causes a race condition with the last aggregation which can result
+        # in a null being returned. For some reason this happens a lot on the GPU in local mode
+        # and not on the CPU in local mode.
+        df.repartition(3).write.mode("overwrite").parquet(spark_tmp_path)
+        df = spark.read.parquet(spark_tmp_path)
+        curDate = df.withColumn("current_date", f.current_date())
+        curDate.createOrReplaceTempView("empty_filter_test_curDate")
+        spark.sql("select current_date, ((select last(current_date) from empty_filter_test_curDate) + interval 1 day) as test from empty_filter_test_curDate").createOrReplaceTempView("empty_filter_test2")
+        return spark.sql(f"select * from empty_filter_test2 where test {op} current_date")
+    assert_gpu_and_cpu_are_equal_collect(do_it)
+
 def test_nondeterministic_filter():
     assert_gpu_and_cpu_are_equal_collect(
             lambda spark : unary_op_df(spark, LongGen(), 1).filter(f.rand(0) > 0.5))

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
@@ -501,10 +501,21 @@ object GpuFilter {
       cb: ColumnarBatch): ColumnarBatch = {
     checkedFilterMask.map { checkedFilterMask =>
       withResource(checkedFilterMask) { checkedFilterMask =>
-        val colTypes = GpuColumnVector.extractTypes(cb)
-        withResource(GpuColumnVector.from(cb)) { tbl =>
-          withResource(tbl.filter(checkedFilterMask)) { filteredData =>
-            GpuColumnVector.from(filteredData, colTypes)
+        if (cb.numCols() <= 0) {
+          val rowCount = withResource(checkedFilterMask.sum(DType.INT32)) { sum =>
+            if (sum.isValid) {
+              sum.getInt
+            } else {
+              0
+            }
+          }
+          new ColumnarBatch(Array(), rowCount)
+        } else {
+          val colTypes = GpuColumnVector.extractTypes(cb)
+          withResource(GpuColumnVector.from(cb)) { tbl =>
+            withResource(tbl.filter(checkedFilterMask)) { filteredData =>
+              GpuColumnVector.from(filteredData, colTypes)
+            }
           }
         }
       }

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala
@@ -198,51 +198,48 @@ case class GpuDateAddInterval(start: Expression,
 
   override def columnarEval(batch: ColumnarBatch): Any = {
 
-    withResourceIfAllowed(left.columnarEval(batch)) { lhs =>
-      withResourceIfAllowed(right.columnarEval(batch)) { rhs =>
-        (lhs, rhs) match {
-          case (l: GpuColumnVector, intvlS: GpuScalar)
-              if intvlS.dataType.isInstanceOf[CalendarIntervalType] =>
-            // Scalar does not support 'CalendarInterval' now, so use
-            // the Scala value instead.
-            // Skip the null check because it wll be detected by the following calls.
-            val intvl = intvlS.getValue.asInstanceOf[CalendarInterval]
+    withResourceIfAllowed(GpuExpressionsUtils.columnarEvalToColumn(left, batch)) { lhs =>
+      withResourceIfAllowed(right.columnarEval(batch)) {
+        case intvlS: GpuScalar if intvlS.dataType.isInstanceOf[CalendarIntervalType] =>
+          // Scalar does not support 'CalendarInterval' now, so use
+          // the Scala value instead.
+          // Skip the null check because it wll be detected by the following calls.
+          val intvl = intvlS.getValue.asInstanceOf[CalendarInterval]
 
-            // ANSI mode checking
-            if(ansiEnabled && intvl.microseconds != 0) {
-              val msg = "IllegalArgumentException: Cannot add hours, minutes or seconds" +
-                  ", milliseconds, microseconds to a date. " +
-                  "If necessary set spark.sql.ansi.enabled to false to bypass this error."
-              throw new IllegalArgumentException(msg)
-            }
+          // ANSI mode checking
+          if (ansiEnabled && intvl.microseconds != 0) {
+            val msg = "IllegalArgumentException: Cannot add hours, minutes or seconds" +
+                ", milliseconds, microseconds to a date. " +
+                "If necessary set spark.sql.ansi.enabled to false to bypass this error."
+            throw new IllegalArgumentException(msg)
+          }
 
-            if (intvl.months != 0) {
-              throw new UnsupportedOperationException("Months aren't supported at the moment")
-            }
-            val microSecToDays = if (intvl.microseconds < 0) {
-              // This is to calculate when subtraction is performed. Need to take into account the
-              // interval( which are less than days). Convert it into days which needs to be
-              // subtracted along with intvl.days(if provided).
-              (intvl.microseconds.abs.toDouble / microSecondsInOneDay).ceil.toInt * -1
-            } else {
-              (intvl.microseconds.toDouble / microSecondsInOneDay).toInt
-            }
-            val daysToAdd = intvl.days + microSecToDays
-            if (daysToAdd != 0) {
-              withResource(Scalar.fromInt(daysToAdd)) { us_s =>
-                withResource(l.getBase.bitCastTo(DType.INT32)) { us =>
-                  withResource(intervalMath(us_s, us)) { intResult =>
-                    GpuColumnVector.from(intResult.castTo(DType.TIMESTAMP_DAYS), dataType)
-                  }
+          if (intvl.months != 0) {
+            throw new UnsupportedOperationException("Months aren't supported at the moment")
+          }
+          val microSecToDays = if (intvl.microseconds < 0) {
+            // This is to calculate when subtraction is performed. Need to take into account the
+            // interval( which are less than days). Convert it into days which needs to be
+            // subtracted along with intvl.days(if provided).
+            (intvl.microseconds.abs.toDouble / microSecondsInOneDay).ceil.toInt * -1
+          } else {
+            (intvl.microseconds.toDouble / microSecondsInOneDay).toInt
+          }
+          val daysToAdd = intvl.days + microSecToDays
+          if (daysToAdd != 0) {
+            withResource(Scalar.fromInt(daysToAdd)) { us_s =>
+              withResource(lhs.getBase.bitCastTo(DType.INT32)) { us =>
+                withResource(intervalMath(us_s, us)) { intResult =>
+                  GpuColumnVector.from(intResult.castTo(DType.TIMESTAMP_DAYS), dataType)
                 }
               }
-            } else {
-              l.incRefCount()
             }
-          case _ =>
-            throw new UnsupportedOperationException("GpuDateAddInterval takes column and " +
-              "interval as an argument only")
-        }
+          } else {
+            lhs.incRefCount()
+          }
+        case _ =>
+          throw new UnsupportedOperationException("GpuDateAddInterval requires a scalar " +
+              "for the interval")
       }
     }
   }

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala
@@ -203,7 +203,7 @@ case class GpuDateAddInterval(start: Expression,
         case intvlS: GpuScalar if intvlS.dataType.isInstanceOf[CalendarIntervalType] =>
           // Scalar does not support 'CalendarInterval' now, so use
           // the Scala value instead.
-          // Skip the null check because it wll be detected by the following calls.
+          // Skip the null check because it will be detected by the following calls.
           val intvl = intvlS.getValue.asInstanceOf[CalendarInterval]
 
           // ANSI mode checking

